trillian/ctfe/ct_server/main.go

```diff
@@ -226,19 +226,22 @@ func main() {
 
 	// Bring up the HTTP server and serve until we get a signal not to.
 	srv := http.Server{Addr: *httpEndpoint, Handler: handler}
+	waitForShutdown := make(chan bool, 1)
 	go awaitSignal(func() {
 		// Allow 60s for any pending requests to finish then terminate any stragglers
 		ctx, cancel := context.WithTimeout(context.Background(), time.Second*60)
 		defer cancel()
 		glog.Info("Shutting down HTTP server...")
 		srv.Shutdown(ctx)
 		glog.Info("HTTP server shutdown")
+		waitForShutdown <- true
 	})
 
 	err = srv.ListenAndServe()
 	if err != http.ErrServerClosed {
 		glog.Warningf("Server exited: %v", err)
 	}
+	<-waitForShutdown
```

Comment:
The problem here is that if `ListenAndServe` returns for a reason other than getting a signal and calling `Shutdown` (can it though?), this wait won't terminate. In case this is possible, a safer approach would be to `WaitGroup.Wait()` here on a group which would have `wg.Add(1)/defer wg.Done()` in line 231.

Reply:
`ListenAndServe` can indeed return other errors, e.g. [net.Listen](https://golang.org/src/net/http/server.go?s=82953:82994#L2707) could fail to start. In this case we probably don't want to wait until the signal is caught, but rather terminate immediately.

Reply:
Ah yeah I see what you mean, a WaitGroup probably is the best solution here then.